pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/BaseResources.java

```diff
@@ -97,6 +98,28 @@ public void set(String path, Function<T, T> modifyFunction) throws MetadataStore
         }
     }
 
+    public void set(String path, OrderedExecutor executor, Function<T, T> modifyFunction) throws MetadataStoreException {
+        CompletableFuture<Void> callback = new CompletableFuture<>();
+        // Use orderedExecutor to ensure that
+        // concurrent updates in each path are handled by a single thread
+        executor.executeOrdered(path, () ->
```

Comment:
There is no more need for this executedOrdered

Reply:
If we remove it, the order cannot be guaranteed. Although there are snapshots now, they are not locked.

Reply:
I am not sure we need to "lock"
now that we know that what is in the case is immutable we are safe

if you need to access the final value of the object we have to return it to the caller and then use the final version.


additionally,  if you use `executeOrdered` you don't know which of the two concurrent operations that are trying to modify the value, so it does not add value.

Reply:
The object instances aren't technically immutable objects where all fields are final. In the "Java Concurrency In Practice" book this type of objects are called "effectively immutable objects". According to this book, they must be safely published when shared to other threads (In the book "3.5.3 Safe publication idioms" and "3.5.4 Effectively Immutable Objects" cover the topic).

An explicity "lock" isn't necessary, but it continues to be necessary to achieve safe publication in a way or another.

Reply:
If roderedExecutor is not used, it will become like this:
Assuming Policies={}
t1: Thread-1 read + clone Policies, Thread-2 read + clone Policies
t2: Thread-1 modifies Policies to {a=1}, Thread-2 modifies Policies to {b=2}
t3: Thread-1 writes back Policies, Thread-2 writes back Policies
Finally Policies={b=2}, the result we expect is {a=1,b=2}

If roderedExecutor is used:
t1: Thread-1 read+clone+modify+update, Policies are {a=1}
t2: Thread-2 read+clone+modify+update, Policies are {a=1,b=2}

Reply:
It won't happen because we are using compare and set while writing to ZK.
We grab the version, we make the mutation and then we writing with the condition that no one else changed the znode in the meantime

Reply:
The version of zk is not stored with Policies, but a version of one broker, which can only ensure that it is safe to update ZK concurrently between different brokers.

t1: Thread-1 writes to zk, updates the local version to v1
t2: Thread-2 writes to zk(use v1) and updates the local version to v2
Local thread safety cannot be guaranteed.

I mentioned a plan in my previous PR, to save a copy of Zk verision to Policies, so that it becomes an optimistic lock. But there are too many changes in this plan.